File: plat/ts/entities/crumblingblock.ts
import * as Assets from 'assets';

import Entity from '../entity.js';
import Dir from '../dir.js';
import CollisionType from '../collisiontype.js';
import { Player } from './player.js';
import World from '../world.js';

class CrumblingBlock extends Entity {
  type = 'CrumblingBlock';
  sprite = Assets.find('crumblingblock');
  collidable = CollisionType.Enabled;
  broken = false;
  breakTime = 0;
  crumbleTime = 40;
  numFrames = 4;

  die() { }

  update(ticks: number, dt: number) {
    if (!this.broken) return;

    this.frame = Math.ceil((ticks - (this.breakTime - this.crumbleTime)) / (this.crumbleTime / this.numFrames));
    if (ticks < this.breakTime) return;

    this.destroyed = true;
    const part = World().spawnDeathParticle(this, 4);
    part.vel[0] = 0;
    part.vel[1] = 0;
  }

  collide(other: Entity, dir: Dir) {
    if (other instanceof Player && dir == Dir.Up && !this.broken) {
      this.broken = true;
      this.breakTime = World().state.ticks + this.crumbleTime;
    }
  }
}

export { CrumblingBlock };

File: csgame/entities/CrumblingBlock.cs
using Slate2D;

[Spawnable]
class CrumblingBlock : Entity
{
    enum Frames
    {
        Uncracked, Cracked1, Cracked2, Cracked3, Cracked4
    }
    Frames[] anim = new[] { Frames.Cracked1, Frames.Cracked2, Frames.Cracked3, Frames.Cracked4 };

    bool Broken = false;
    uint BreakTime = 0;
    const uint CrumbleTime = 40;

    public override void Die() { }

    public CrumblingBlock(LDTKEntity ent) : base(ent)
    {
        Sprite = Assets.Find("crumblingblock");
        Collidable = CollisionType.Enabled;
        Layer = Layer.Background;

    }

    public override void Update(uint ticks, float dt)
    {
        if (!Broken) return;

        Frame = Animate(anim, BreakTime - CrumbleTime, BreakTime);
        if (Ticks < BreakTime) return;

        Destroyed = true;
        var part = Main.World.SpawnDeathParticle(this, 4);
        part.Vel = (0, 0);
    }

    public override void Collide(Entity other, Dir dir)
    {
        if (other is Player && dir == Dir.Up && !Broken) {
            Broken = true;
            BreakTime = Ticks + CrumbleTime;
        }
    }
}